### PR TITLE
Feature/svm grouping

### PR DESCRIPTION
**Pull request by mktal**
Add grouping support for SVM regression

Issue: https://issues.apache.org/jira/browse/MADLIB-913

### CODE REVIEW

File: src/modules/convex/linear_svm_igd.cpp [hunk 1]
```diff
@@ -95,19 +90,16 @@ linear_svm_igd_transition::run(AnyType &args) {
...
     // apply IGD with regularization
-    if (isL2) {
-        L2<GLMModel>::scaling(state.algo.incrModel, lambda, nTuples, state.task.stepsize);
-        LinearSVMIGDAlgorithm::transition(state, tuple);
-    } else {
-        LinearSVMIGDAlgorithm::transition(state, tuple);
-        L1<GLMModel>::clipping(state.algo.incrModel, lambda, nTuples, state.task.stepsize);
-    }
-    // objective function and its gradient
+    L2<GLMModel>::scaling(state.task.model, state.algo.incrModel, state.task.stepsize);
```
Comment (iyerr3) [inline src/modules/convex/linear_svm_igd.cpp]: Why not check the isL2 variable and only perform the necessary computation?

File: src/modules/convex/linear_svm_igd.cpp [hunk 2]
```diff
@@ -152,9 +144,18 @@ linear_svm_igd_final::run(AnyType &args) {
 
     // Aggregates that haven't seen any data just return Null.
     if (state.algo.numRows == 0) { return Null(); }
+
```
Comment (iyerr3) [inline src/modules/convex/linear_svm_igd.cpp]: Same as above - I would prefer to have one `if (isL2)` that decides between the L2 and L1 functions.

File: src/modules/convex/linear_svm_igd.cpp (hunk 1 above)
Comment (mktal) [inline src/modules/convex/linear_svm_igd.cpp]: The L2 and L1 regularization are not exclusive. We can perform both of them by setting the corresponding regularization parameters to non-zero.

File: src/ports/postgres/modules/utilities/in_mem_group_control.py_in [hunk 3]
```diff
@@ -7,6 +7,141 @@
...
+            return failed_grp_keys
+        # initialize state to None
+        if col_grp_state == '':
+            self._is_none = True
+            for s in ret_states:
+                self._state[s[col_grp_key]] = None
+            return failed_grp_keys
+        for t in ret_states:
+            _grp_key, _grp_state = t[col_grp_key], t[col_grp_state]
+            if _grp_state is None: failed_grp_keys.append(_grp_key)
```
Comment (iyerr3) [inline src/ports/postgres/modules/utilities/in_mem_group_control.py_in]: Let's put the statement in a separate line per [Pep 8](https://www.python.org/dev/peps/pep-0008/#other-recommendations)

File: src/modules/convex/linear_svm_igd.cpp (hunk 1 above)
Comment (mktal) [inline src/modules/convex/linear_svm_igd.cpp]: For now the users can only choose either l1 or l2. We can support both in the future if necessary

File: CMakeLists.txt [hunk 4]
```diff
@@ -103,6 +103,8 @@ if(CMAKE_COMPILER_IS_GNUCXX)
     if(APPLE)
         set(CMAKE_INCLUDE_SYSTEM_FLAG_CXX "-isystem ")
     endif(APPLE)
+elseif(CMAKE_C_COMPILER_ID MATCHES "Clang")
```
Comment (iyerr3) [inline CMakeLists.txt]: @mktal let's move this out as a separate JIRA+PR since it's not related to SVM.

**Pull request closed without merging**